src/viam/trajex/types/epsilon.hpp

```diff
@@ -77,10 +81,91 @@ class epsilon {
         return epsilon{-value_};
     }
 
+    template <typename T>
+    class wrapper;
+
+    template <typename T>
+    constexpr wrapper<T> wrap(const T& t) const noexcept;
+
    private:
     double value_;
 };
 
+template <typename T>
+class epsilon::wrapper {
+    explicit constexpr wrapper(const epsilon& e, T const& t) : e_{e}, t_{t} {}
+
+    friend class epsilon;
+
+    // clang-format off
+    template <typename U, typename V>
+    friend constexpr auto operator<=>(const wrapper<U>& lhs, const wrapper<V>& rhs) noexcept
+        requires std::three_way_comparable_with<U, V> && epsilon_details::explicitly_convertible_to_double<U> && epsilon_details::explicitly_convertible_to_double<V>;
+
+    template <typename U, typename V>
+    friend constexpr bool operator==(const wrapper<U>& lhs, const wrapper<V>& rhs) noexcept
+        requires std::three_way_comparable_with<U, V> && epsilon_details::explicitly_convertible_to_double<U> && epsilon_details::explicitly_convertible_to_double<V>;
+    // clang-format on
+
+   private:
+    const epsilon& e_;
+    T const& t_;
+};
+
+template <typename T>
+constexpr epsilon::wrapper<T> epsilon::wrap(const T& t) const noexcept {
+    return wrapper<T>(*this, t);
+}
+
+// clang-format off
+
+///
+/// Three-way comparison between wrapped values with epsilon tolerance.
+///
+/// Compares two values accounting for epsilon tolerance. Values are considered
+/// equivalent if their difference is within tolerance, otherwise ordered by
+/// which value is significantly greater.
+///
+/// When wrappers have different epsilon values, uses the stricter (minimum) tolerance.
+///
+/// @param lhs Left wrapped value
+/// @param rhs Right wrapped value
+/// @return weak_ordering::less if rhs significantly exceeds lhs,
+///         weak_ordering::greater if lhs significantly exceeds rhs,
+///         weak_ordering::equivalent if values are within tolerance
+///
+template <typename T, typename U>
+constexpr auto operator<=>(const epsilon::wrapper<T>& lhs, const epsilon::wrapper<U>& rhs) noexcept
+    requires std::three_way_comparable_with<T, U> && epsilon_details::explicitly_convertible_to_double<T> && epsilon_details::explicitly_convertible_to_double<U> {
+    const double tol = std::min(static_cast<double>(lhs.e_), static_cast<double>(rhs.e_));
```

Comment:
Alternatively, could throw if `&lhs.e_ != &rhs.e_`, but I didn't want exceptions in a hot path, and comparing numbers definitely is that. Personally, I think most of the time this check will boil off because the compiler will be able to see that `e.wrap(foo) < e.wrap(bar)` has the same `e` in both places. Hopefully.

Reply:
I think this is fine as is.